qos/evm/state.go

```diff
@@ -29,72 +31,164 @@ type ServiceState struct {
 			continue
 		}
 
+		// Update the perceived block number.
 		s.perceivedBlockNumber = blockNumber
 
-		logger.With("endpoint_block_number", blockNumber).Info().Msg("Updating latest block height")
+		// Attempt to retrieve the archival balance from the endpoint.
+		balance, err := endpoint.getArchivalBalance()
+		if err != nil {
+			logger.Info().Err(err).Msg("Skipping endpoint without archival balance")
+			continue
+		}
+
+		// Only count non-empty balances toward consensus.
+		if s.archivalState.balance == "" && balance != "" {
+			s.archivalState.balanceConsensus[balance]++
+		}
+	}
+
+	// Update archival block number and archival balance only if not yet set.
+	if s.perceivedBlockNumber != 0 && s.archivalState.blockNumberHex == "" {
+		s.assignArchivalBlockNumber()
+	}
+	if s.archivalState.balance == "" {
+		s.updateArchivalBalance(consensusThreshold)
 	}
 
 	return nil
 }
 
-// validateEndpointBlockNumber validates the supplied endpoint against the supplied perceived block number for the EVM blockchain.
-func validateEndpointBlockNumber(endpoint endpoint, perceivedBlockNumber uint64) error {
-	blockNumber, err := endpoint.GetBlockNumber()
-	if err != nil {
-		return err
+// assignArchivalBlockNumber returns a random archival block number based on the perceived block number.
+func (s *ServiceState) assignArchivalBlockNumber() string {
+	archivalThreshold := s.archivalCheckConfig.Threshold
+	minArchivalBlock := s.archivalCheckConfig.ContractStartBlock
+
+	var result string
+	if s.perceivedBlockNumber <= archivalThreshold {
+		result = blockNumberToHex(0)
+	} else {
+		maxBlockNumber := s.perceivedBlockNumber - archivalThreshold
+		if maxBlockNumber < minArchivalBlock {
+			result = blockNumberToHex(minArchivalBlock)
+		} else {
+			r := rand.New(rand.NewSource(time.Now().UnixNano()))
+			rangeSize := maxBlockNumber - minArchivalBlock + 1
+			result = blockNumberToHex(minArchivalBlock + (r.Uint64() % rangeSize))
+		}
 	}
 
-	if blockNumber < perceivedBlockNumber {
-		return fmt.Errorf("endpoint has block height %d, perceived block height is %d", blockNumber, perceivedBlockNumber)
+	s.archivalState.blockNumberHex = result
+	return result
+}
+
+// updateArchivalBalance checks for consensus and updates the archival balance if it hasn't been set yet.
+func (s *ServiceState) updateArchivalBalance(consensusThreshold int) {
+	for balance, count := range s.archivalState.balanceConsensus {
+		if count >= consensusThreshold {
+			s.archivalState.balance = balance
+			// Reset consensus map after consensus is reached.
+			s.archivalState.balanceConsensus = make(map[string]int)
+			break
+		}
 	}
+}
 
-	return nil
+// shouldPerformArchivalCheck returns a boolean indicating whether the archival check should be performs.
+// This depends on:
+//   - The archival check is enabled for the service
+//   - The archival block number to check the balance of has been set in the service state.
```

Comment:
Please add a why/when to make it even clearer!

```suggestion
//   - The archival check is enabled for the service in the configs
//   - The archival block number to check the balance of has been set in the service state during ????
```